homeassistant/components/sensor/roomba.py

```diff
@@ -0,0 +1,171 @@
+            charging = phase_data == 'charge' if phase_data else None
+            self._state = roomba_data.get('batPct', None)
+            self._attrs = {ATTR_BATTERY_CHARGING: charging}
+            if roomba_name:
+                self._sensor_name = '{} Battery'.format(roomba_name)
+        elif self.sensor_type == SENSOR_TYPE_POSITION:
+            position_data = roomba_data.get('pose', None)
+            pos_x = position_data.get('point', {}).get('x', None)
+            pos_y = position_data.get('point', {}).get('y', None)
+            theta = position_data.get('theta', None)
+            if all(item is not None for item in [pos_x, pos_y, theta]):
+                self._state = '({}, {}, {})'.format(pos_x, pos_y, theta)
+            else:
+                self._state = None
+            self._attrs = {
+                ATTR_POSITION_X: pos_x,
+                ATTR_POSITION_Y: pos_y,
+                ATTR_POSITION_THETA: theta
+            }
+            if roomba_name:
+                self._sensor_name = '{} Position'.format(roomba_name)
+        elif self.sensor_type == SENSOR_TYPE_STATUS:
+            clean_mission_status = roomba_data.get('cleanMissionStatus', {})
+            phase = clean_mission_status.get('phase', None)
+            if phase in PHASES:
+                self._state = PHASES.get(phase)
+            else:
+                # Unknown phase. Use as is.
+                self._state = phase
+            if roomba_name:
+                self._sensor_name = '{} Status'.format(roomba_name)
+
+    def update(self):
+        """Update the properties of sensor."""
+        _LOGGER.debug('Update of Roomba %s sensor', self.sensor_type)
+        self.roomba_hub.update()
+        self.__set_sensor_state_from_hub()
+        _LOGGER.debug('%s sensor state: %s', self.sensor_type, self._state)
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        if self.sensor_type == SENSOR_TYPE_BATTERY:
+            if self._state is None:
+                return 'mdi:battery-unknown'
+            rounded_level = round(self._state, -1)
+            charging = self._attrs.get(ATTR_BATTERY_CHARGING, False)
+            if charging:
+                # Why is MDI missing 10, 50, 70?
+                if rounded_level == 10:
+                    return 'mdi:battery-charging-20'
+                elif rounded_level == 50:
+                    return 'mdi:battery-charging-60'
+                elif rounded_level == 70:
+                    return 'mdi:battery-charging-80'
+                else:
+                    return 'mdi:battery-charging-{}'.format(rounded_level)
+            else:
+                if rounded_level < 10:
+                    return 'mdi:battery-outline'
+                elif rounded_level == 100:
+                    return 'mdi:battery'
+                else:
+                    return 'mdi:battery-{}'.format(rounded_level)
+            return 'mdi:battery'
```

Comment:
Yes, that was also my opinion. A cool helper function inside utils would be the best.